org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java

```diff
@@ -0,0 +1,239 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonarlint.eclipse.core.internal.tracking;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class Tracker<RAW extends Trackable, BASE extends Trackable> {
+
+  public Tracking<RAW, BASE> track(Input<RAW> rawInput, Input<BASE> baseInput) {
+    Tracking<RAW, BASE> tracking = new Tracking<>(rawInput, baseInput);
+
+    // 1. match issues with same rule, same line and same line hash, but not necessarily with same message
+    match(tracking, LineAndLineHashKeyFactory.INSTANCE);
+
+    // 3. match issues with same rule, same message and same line hash
+    match(tracking, LineHashAndMessageKeyFactory.INSTANCE);
+
+    // 4. match issues with same rule, same line and same message
+    match(tracking, LineAndMessageKeyFactory.INSTANCE);
+
+    // 5. match issues with same rule and same line hash but different line and different message.
+    // See SONAR-2812
+    match(tracking, LineHashKeyFactory.INSTANCE);
+
+    return tracking;
+  }
+
+  private void match(Tracking<RAW, BASE> tracking, SearchKeyFactory factory) {
+    if (tracking.isComplete()) {
+      return;
+    }
+
+    Map<SearchKey, List<BASE>> baseSearch = new HashMap<>();
+    for (BASE base : tracking.getUnmatchedBases()) {
+      SearchKey searchKey = factory.create(base);
+      if (!baseSearch.containsKey(searchKey)) {
+        baseSearch.put(searchKey, new ArrayList<BASE>());
+      }
+      baseSearch.get(searchKey).add(base);
+    }
+
+    for (RAW raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.create(raw);
+      Collection<BASE> bases = baseSearch.get(rawKey);
+      if (bases != null && !bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
+        // Message could be checked to take the best one.
+        BASE match = bases.iterator().next();
+        tracking.match(raw, match);
+        baseSearch.get(rawKey).remove(match);
+      }
+    }
+  }
+
+  private interface SearchKey {
+  }
+
+  private interface SearchKeyFactory {
+    SearchKey create(Trackable trackable);
+  }
+
+  private static class LineAndLineHashKey implements SearchKey {
+    private final String ruleKey;
+    private final Integer lineHash;
+    private final Integer line;
+
+    LineAndLineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.line = trackable.getLine();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a type test to this method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)